File: streampark-flink/streampark-flink-packer/src/main/scala/org/apache/streampark/flink/packer/pipeline/impl/FlinkRemoteBuildPipeline.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.packer.pipeline.impl

import org.apache.streampark.common.enums.FlinkJobType
import org.apache.streampark.common.fs.{FsOperator, LfsOperator}
import org.apache.streampark.common.util.Implicits._
import org.apache.streampark.flink.packer.maven.MavenTool
import org.apache.streampark.flink.packer.pipeline._

import java.io.File

/** Building pipeline for flink standalone session mode */
class FlinkRemoteBuildPipeline(request: FlinkRemotePerJobBuildRequest) extends BuildPipeline {

  override def pipeType: PipelineTypeEnum = PipelineTypeEnum.FLINK_STANDALONE

  override def offerBuildParam: FlinkRemotePerJobBuildRequest = request

  /** The construction logic needs to be implemented by subclasses */
  @throws[Throwable]
  override protected def buildProcess(): ShadedBuildResponse = {
    // create workspace.
    // the sub workspace path like: APP_WORKSPACE/jobName
    if (request.skipBuild) {
      ShadedBuildResponse(request.workspace, request.customFlinkUserJar)
    } else {
      execStep(1) {
        LfsOperator.mkCleanDirs(request.workspace)
        logInfo(s"Recreate building workspace: ${request.workspace}")
      }.getOrElse(throw getError.exception)
      // build flink job shaded jar
      val shadedJar =
        execStep(2) {
          request.flinkJobType match {
            case FlinkJobType.FLINK_SQL =>
              val output = MavenTool.buildFatJar(
                request.mainClass,
                request.providedLibs,
                request.getShadedJarPath(request.workspace))
              logInfo(s"output shaded flink job jar: ${output.getAbsolutePath}")
              output
            case _ => new File(request.customFlinkUserJar);
          }
        }.getOrElse(throw getError.exception)

      val mavenJars =
        execStep(3) {
          request.flinkJobType match {
            case FlinkJobType.PYFLINK =>
              val mavenArts =
                MavenTool.resolveArtifacts(request.dependencyInfo.mavenArts)
              mavenArts.map(_.getAbsolutePath) ++ request.dependencyInfo.extJarLibs
            case _ => List[String]()
          }
        }.getOrElse(throw getError.exception)

      execStep(4) {
        request.flinkJobType match {
          case FlinkJobType.PYFLINK =>
            mavenJars.foreach(jar => {
              val lfs: FsOperator = FsOperator.lfs
              val lib = request.workspace.concat("/lib")
              lfs.mkdirsIfNotExists(lib)
              val originFile = new File(jar)
              if (originFile.isFile) {
                lfs.copy(originFile.getAbsolutePath, lib)
              }

            })
          case _ =>
        }
      }.getOrElse(throw getError.exception)
      ShadedBuildResponse(request.workspace, shadedJar.getAbsolutePath)
    }

  }

}

object FlinkRemoteBuildPipeline {
  def of(request: FlinkRemotePerJobBuildRequest): FlinkRemoteBuildPipeline =
    new FlinkRemoteBuildPipeline(request)
}

File: streampark-flink/streampark-flink-packer/src/main/java/org/apache/streampark/flink/packer/pipeline/impl/FlinkRemoteBuildPipeline.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.packer.pipeline.impl;

import org.apache.streampark.common.enums.FlinkJobType;
import org.apache.streampark.common.fs.FsOperator;
import org.apache.streampark.common.fs.LfsOperator;
import org.apache.streampark.flink.packer.maven.MavenTool;
import org.apache.streampark.flink.packer.pipeline.BuildPipeline;
import org.apache.streampark.flink.packer.pipeline.FlinkRemotePerJobBuildRequest;
import org.apache.streampark.flink.packer.pipeline.PipelineTypeEnum;
import org.apache.streampark.flink.packer.pipeline.ShadedBuildResponse;

import java.io.File;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/** Building pipeline for flink standalone session mode */
public class FlinkRemoteBuildPipeline extends BuildPipeline {

    private final FlinkRemotePerJobBuildRequest request;

    public FlinkRemoteBuildPipeline(FlinkRemotePerJobBuildRequest request) {
        this.request = request;
    }

    @Override
    public PipelineTypeEnum pipeType() {
        return PipelineTypeEnum.FLINK_STANDALONE;
    }

    @Override
    public FlinkRemotePerJobBuildRequest offerBuildParam() {
        return request;
    }

    @Override
    public ShadedBuildResponse buildProcess() {
        if (request.skipBuild()) {
            return new ShadedBuildResponse(request.workspace(), request.customFlinkUserJar());
        }

        execStep(1, () -> {
            LfsOperator.mkCleanDirs(request.workspace());
            logInfo("Recreate building workspace: " + request.workspace());
            return null;
        });

        File shadedJar =
            execStep(
                2,
                () -> {
                    if (request.flinkJobType() == FlinkJobType.FLINK_SQL) {
                        File output =
                            MavenTool.buildFatJar(
                                request.mainClass(),
                                request.providedLibs(),
                                request.getShadedJarPath(request.workspace()));
                        logInfo("output shaded flink job jar: " + output.getAbsolutePath());
                        return output;
                    }
                    return new File(request.customFlinkUserJar());
                });

        List<String> mavenJars =
            execStep(
                3,
                () -> {
                    if (request.flinkJobType() == FlinkJobType.PYFLINK) {
                        List<File> mavenArts =
                            MavenTool.resolveArtifacts(request.dependencyInfo().mavenArts());
                        List<String> paths =
                            mavenArts.stream()
                                .map(File::getAbsolutePath)
                                .collect(Collectors.toList());
                        paths.addAll(request.dependencyInfo().extJarLibs());
                        return paths;
                    }
                    return Collections.<String>emptyList();
                });

        execStep(
            4,
            () -> {
                if (request.flinkJobType() == FlinkJobType.PYFLINK) {
                    FsOperator lfs = FsOperator.lfs();
                    String lib = request.workspace().concat("/lib");
                    lfs.mkdirsIfNotExists(lib);
                    for (String jar : mavenJars) {
                        File originFile = new File(jar);
                        if (originFile.isFile()) {
                            lfs.copy(originFile.getAbsolutePath(), lib);
                        }
                    }
                }
                return null;
            });

        return new ShadedBuildResponse(request.workspace(), shadedJar.getAbsolutePath());
    }

    public static FlinkRemoteBuildPipeline of(FlinkRemotePerJobBuildRequest request) {
        return new FlinkRemoteBuildPipeline(request);
    }
}
